File: docs/test-cases/redos.js
// ReDoS — intentionally vulnerable. DO NOT deploy.

// BUG: nested quantifiers cause catastrophic backtracking
const EMAIL_RE = /^([a-zA-Z0-9_.+-]+)+@([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$/;

// BUG: overlapping alternation
const SLUG_RE = /^([\w-]+)*$/;

function validateEmail(input) {
  return EMAIL_RE.test(input); // hangs on "aaa...@" with no domain
}

function validateSlug(input) {
  return SLUG_RE.test(input); // hangs on "aaa...!"
}

module.exports = { validateEmail, validateSlug };

File: docs/test-cases/redos.py
"""ReDoS — intentionally vulnerable. DO NOT deploy."""
import re

# BUG: nested quantifiers cause catastrophic backtracking
EMAIL_RE = re.compile(r"^([a-zA-Z0-9_.+-]+)+@([a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$")

def validate_email(email: str) -> bool:
    return bool(EMAIL_RE.match(email))

# BUG: overlapping alternation with repetition
SLUG_RE = re.compile(r"^([\w-]+)*$")

def validate_slug(slug: str) -> bool:
    return bool(SLUG_RE.match(slug))
